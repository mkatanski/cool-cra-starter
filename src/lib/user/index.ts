export * from "./userContexts";
export * from "./userProvider";
export * from "./userReducer";
export * from "./userActions";
export * from "./userState";
