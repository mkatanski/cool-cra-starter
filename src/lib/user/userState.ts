export type UserState = {
  isLoggedIn: boolean;
  username: string;
  accessToken?: string;
  refreshToken?: string;
};

export const defaultState: UserState = {
  isLoggedIn: true,
  username: "",
};
