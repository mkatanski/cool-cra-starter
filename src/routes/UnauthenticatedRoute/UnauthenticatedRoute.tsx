import React, { useContext, ComponentType } from "react";
import {
  Route,
  RouteProps,
  Redirect,
  RouteComponentProps,
} from "react-router-dom";

import { UserStateContext } from "lib/user";
import { paths } from "config/paths";

export interface UnauthenticatedRouteProps extends RouteProps {
  component: ComponentType<RouteComponentProps>;
  allowAuthenticated?: boolean;
}

/** Route available only if user is unauthenticated */
export const UnauthenticatedRoute: React.FC<UnauthenticatedRouteProps> = ({
  component: Component,
  allowAuthenticated,
  ...rest
}: UnauthenticatedRouteProps) => {
  const { isLoggedIn } = useContext(UserStateContext);

  return isLoggedIn && !allowAuthenticated ? (
    <Redirect to={paths.dashboard} push />
  ) : (
    <Route {...rest} render={(matchProps) => <Component {...matchProps} />} />
  );
};
