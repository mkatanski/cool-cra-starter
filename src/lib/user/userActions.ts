import { UserState } from "./userState";

export type SetUserAction = {
  type: "SET_USER";
  user: UserState;
};

export type UpdateUserAction = {
  type: "UPDATE_USER";
  user: Partial<UserState>;
};

export type UserAction = SetUserAction | UpdateUserAction;

export const clearUserData = (): SetUserAction => ({
  type: "SET_USER",
  user: {
    isLoggedIn: false,
    username: "",
  },
});

export const setUserData = (username: string): SetUserAction => ({
  type: "SET_USER",
  user: {
    isLoggedIn: true,
    username,
  },
});
