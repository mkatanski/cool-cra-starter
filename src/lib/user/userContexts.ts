import { UserState, defaultState } from "./userState";
import { UserAction } from "./userActions";
import { makeStoreContexts } from "utils/store";

const [StateContext, DispatchContext] = makeStoreContexts<
  UserState,
  UserAction
>(defaultState);

export const UserStateContext = StateContext;
export const UserDispatchContext = DispatchContext;
