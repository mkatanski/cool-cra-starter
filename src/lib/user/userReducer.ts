import { UserAction } from "./userActions";
import { UserState } from "./userState";

export const userReducer = (
  state: UserState,
  action: UserAction
): UserState => {
  switch (action.type) {
    case "SET_USER":
      return action.user;
    case "UPDATE_USER":
      return {
        ...state,
        ...action.user,
      };
  }
};
