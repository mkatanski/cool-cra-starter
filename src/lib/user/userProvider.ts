import { UserStateContext, UserDispatchContext } from "./userContexts";
import { userReducer } from "./userReducer";
import { makeStoreProvider } from "utils/store";
import { defaultState } from "./userState";

export const UserProvider = makeStoreProvider({
  StateContext: UserStateContext,
  DispatchContext: UserDispatchContext,
  reducer: userReducer,
  defaultState,
});
