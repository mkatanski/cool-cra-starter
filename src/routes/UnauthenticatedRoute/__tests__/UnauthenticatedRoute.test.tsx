import React from "react";
import TestRenderer from "react-test-renderer";

import { UnauthenticatedRoute } from "../UnauthenticatedRoute";
import { UserStateContext } from "lib/user";

const TestComponent = () => <div>TestComponent</div>;

describe("routes", () => {
  describe("UnauthenticatedRoute", () => {
    it("should render UnauthenticatedRoute component with default context", () => {
      const tree = TestRenderer.create(
        <UnauthenticatedRoute component={TestComponent} />
      );
      expect(tree).toMatchSnapshot();
    });

    it("should render redirect to dashboard if user is authenticated", () => {
      const tree = TestRenderer.create(
        <UserStateContext.Provider
          value={{ isLoggedIn: true, username: "test" }}
        >
          <UnauthenticatedRoute component={TestComponent} />
        </UserStateContext.Provider>
      );

      expect(tree).toMatchSnapshot();
    });
  });
});
