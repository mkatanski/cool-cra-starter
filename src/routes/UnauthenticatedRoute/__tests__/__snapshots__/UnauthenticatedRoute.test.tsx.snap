// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`routes UnauthenticatedRoute should render UnauthenticatedRoute component with default context 1`] = `
<div
  id="ReactRouter.Route"
>
  <div>
    TestComponent
  </div>
</div>
`;

exports[`routes UnauthenticatedRoute should render redirect to dashboard if user is authenticated 1`] = `
<div
  id="ReactRouter.Redirect"
  push={true}
  to="/dashboard"
/>
`;
